### DISCUSSION TITLE
Sort columns without splitting lower levels

### DISCUSSION THREAD

**Discussion by lingyixu** [DingWB/PyComplexHeatmap#84] [Q&A] [answered]
First, thanks for making and maintaining this package for Python users.

In the context of sorting multilevel columns, i.e. passing a dataframe to `col_split`, would it be possible to split columns on the first level only and keep lower level ones unsplit? An example could be an extension of [this Clustermap](https://dingwb.github.io/PyComplexHeatmap/build/html/notebooks/advanced_usage.html#cluster_within_groups-&&-cluster_between_groups:-col_split=*,-col_split_order=%22cluster_between_groups%22,col_cluster=True) in Advanced Usage. When there're multiple `AB` values under a group, or multiple `Groups` values under one `AB`, passing both `AB` and `Groups` to `col_split` and specifying a non-zero `col_split_gap` will result in gaps between every adjacent `AB` and `Groups`. This is sometimes not ideal for identifying high-level groups. On the other hand, passing one single level to `col_split` doesn't necessarily guarantee a consistent order of lower levels across the high level. I'd appreciate it if an option could be provided for specifying whether to split a sorted level. Also, please kindly point me to any reference if this function is already incorporated within the package.

Thanks again and happy to chat more.

**Comment by DingWB** [comment, marked as answer]
Hi @lingyixu 
Thanks for your feedback. 
To generate the heatmap you described, you can sort the dataframe by lower level first and then set `col_split =upper level` and `col_cluster=False`. 
Do you have any suggestions for the parameter? For example, if `col_split=df_col.loc[:,['A','B']]` and you don't need the gap between groups in B (lower level),  what parameter should I add? Please give me your idea.

**Comment by lingyixu** [comment]
Hi @DingWB , thanks for getting back.

Passing a pre-sorted dataframe as you described is a good workaround. To my mind, it's worth documenting this with a line or two.

Speaking of the extra parameter, my thought was to bring it into [`plot_matrix()`](https://github.com/DingWB/PyComplexHeatmap/blob/f5dd95c3ab361fafd238dad199a712fff1e655fb/PyComplexHeatmap/clustermap.py#L1949) function and to use it along with `col_order`. It would be great to allow `col_order` to support multiple lists, such as taking in a dict in the form of `{col_name:col_order}`. The new param would then specify which column(s) to split, such as taking in a list in the form of `[col_name1, col_name2, ...]`, with column names consistent with the keys in `col_order`. This way, `ncols = len(col_order)` could be updated with `ncols = len(col_order[col_name])`, and `self.wspace` will be updated accordingly. Does this make sense?

So far this is what I have in mind. I didn't get a chance to verify whether this is valid or not, so I could be biased. Please feel free to let me know what you think. Thanks.

**Comment by DingWB** [comment]
Hi @lingyixu ,

Thanks for your suggestion. I will think about it.
For the documentation, I do have a sentence [here](https://dingwb.github.io/PyComplexHeatmap/build/html/notebooks/advanced_usage.html#cluster_between_groups:-col_split=*,-col_split_order=%22cluster_between_groups%22,col_cluster=False), maybe it is not very obvious.
Please don't hesitate to update the documentation (if something is missing) under the notebooks directory and make a pull request.

**Comment by lingyixu** [comment]
Hi @DingWB ,

Thanks for pointing that out. I see your notes there, appreciate it.

May I suggest adding a few words in the docstring, for instance, "setting it to False will preserve the column order of the input data", given that `col_cluster` it's set to True by default. Same for `row_cluster`. If you don't mind, I can proceed with making the edits and creating a pull request.

https://github.com/DingWB/PyComplexHeatmap/blob/f5dd95c3ab361fafd238dad199a712fff1e655fb/PyComplexHeatmap/clustermap.py#L1016-L1017

Thanks again for all your comments and feedback. For now, I'm going to mark this discussion as answered. Please feel free to add any further comments in this thread, and I’ll check back to see if any additional input is needed from my side.

**Comment by DingWB** [comment]
Hi @lingyixu ,

Thanks for your suggestion, please go ahead and add whatever you want and make a pull request.
